include/seqan3/alignment/matrix/debug_matrix.hpp

```diff
@@ -0,0 +1,391 @@
+    debug_matrix & operator=(debug_matrix const &) = default; //!< Defaulted
+    debug_matrix & operator=(debug_matrix &&) = default; //!< Defaulted
+    ~debug_matrix() = default;  //!< Defaulted
+
+    /*!\brief Construct the matrix out of the *entries*, the *rows*, and the *cols*.
+     * \param entries The entry values as a flat std::vector <#entry_type>.
+     * \param rows    The number of rows.
+     * \param cols    The number of columns.
+     */
+    debug_matrix(std::vector<entry_type> entries, size_t const rows, size_t const cols)
+        : _matrix{std::move(entries), rows, cols},
+          _sequence1{std::nullopt},
+          _sequence2{std::nullopt}
+    {}
+
+    /*!\brief Construct the matrix out of the *entries*, the *rows*, and the *cols* and the corresponding sequences.
+     * \param entries The entry values as a flat std::vector <#entry_type>.
+     * \param rows    The number of rows.
+     * \param cols    The number of columns.
+     * \param sequence1 The first sequence of the sequence alignment.
+     * \param sequence2 The second sequence of the sequence alignment.
+     */
+    debug_matrix(std::vector<entry_type> entries, size_t const rows, size_t const cols,
+                 sequence1_t sequence1, sequence2_t sequence2)
+        : _matrix{std::move(entries), rows, cols},
+          _sequence1{std::forward<sequence1_t>(sequence1)},
+          _sequence2{std::forward<sequence2_t>(sequence2)}
+    {
+        assert(cols == _sequence1.size() + 1u);
+        assert(rows == _sequence2.size() + 1u);
+    }
+
+    /*!\brief Construct the matrix out of an existing matrix.
+     * \param matrix An alignment matrix; Must model seqan3::detail::Matrix.
+     */
+    debug_matrix(matrix_t matrix)
+        : _matrix{std::forward<matrix_t>(matrix)},
+          _sequence1{std::nullopt},
+          _sequence2{std::nullopt}
+    {}
+
+    /*!\brief Construct the matrix out of an existing matrix and two sequences.
+     * \param matrix An alignment matrix; Must model seqan3::detail::Matrix.
+     * \param sequence1 The first sequence of the sequence alignment.
+     * \param sequence2 The second sequence of the sequence alignment.
+     */
+    debug_matrix(matrix_t matrix, sequence1_t sequence1, sequence2_t sequence2)
+        : _matrix{std::forward<matrix_t>(matrix)},
+          _sequence1{std::forward<sequence1_t>(sequence1)},
+          _sequence2{std::forward<sequence2_t>(sequence2)}
+    {
+        assert(_matrix.cols() == _sequence1.size() + 1u);
+        assert(_matrix.rows() == _sequence2.size() + 1u);
+    }
+    //!\}
+
+    //!\copydoc seqan3::detail::Matrix::rows
+    size_t rows() const noexcept
+    {
+        return _matrix.rows();
+    }
+
+    //!\copydoc seqan3::detail::Matrix::cols
+    size_t cols() const noexcept
+    {
+        return _matrix.cols();
+    }
+
+    //!\copydoc seqan3::detail::debug_matrix::_sequence1
+    sequence1_t const & sequence1() const noexcept
```

Comment:
In the remaining part of the alignment we used `first_sequence` and `second_sequence` so far. It is not super important, but it would be altogether be better to keep the naming scheme consistent.